lib/src/repo.rs

```diff
@@ -810,12 +810,42 @@ impl MutableRepo {
         Ok(commit)
     }
 
-    /// Record a commit as having been rewritten in this transaction. This
-    /// record is used by `rebase_descendants()`.
+    /// Record a commit as having been rewritten to one or more ids in this
+    /// transaction.
     ///
-    /// Rewritten commits don't have to be recorded here. This is just a
-    /// convenient place to record it. It won't matter after the transaction
-    /// has been committed.
+    /// This record is used by `rebase_descendants` to know which commits have
+    /// children that need to be rebased, and where to rebase them to. See the
+    /// docstring for `record_rewritten_commit` for details.
+    //  TODO(ilyagr): Make this API saner, e.g. make `self.rewritten_commits` public
+    // and make empty values correspond to abandoned commits.
+    pub fn set_rewritten_commit(
+        &mut self,
+        old_id: CommitId,
+        new_ids: impl IntoIterator<Item = CommitId>,
+    ) {
+        assert_ne!(old_id, *self.store().root_commit_id());
+        self.rewritten_commits
+            .insert(old_id, new_ids.into_iter().collect());
+    }
+
+    /// Record a commit as having been rewritten in this transaction. If it was
+    /// already rewritten, mark it as divergent (unlike `set_rewritten_commit`)
+    ///
+    /// This record is used by `rebase_descendants` to know which commits have
+    /// children that need to be rebased, and where to rebase the children (as
+    /// well as branches) to.
+    ///
+    /// The `rebase_descendants` logic treats these records as follows:
+    ///
+    /// - If a commit is recorded as rewritten to a single commit, its
+    ///   descendants would be rebased to become descendants of `new_id`. Any
+    ///   branches at `old_id` are also moved to `new_id`.
+    /// - If a commit is recorded as rewritten to more than one commit, it is
+    ///   assumed to have become divergent. Its descendants are *not* rebased.
+    ///   TODO(ilyagr): Test and check what happens to branches in this case.
```

Comment:
They should become divergent. There's a test case here: https://github.com/martinvonz/jj/blob/320d15412b0326b96e69e00421e465f6bb9b49b8/lib/tests/test_rewrite.rs#L1108-L1175, but there are no descendants in that test case. I think you can update the text here to say that that's what should happen. If it turns out to not work when there are descendants, we should probably consider that a bug.

Reply:
Thanks! I updated the comment and also added a TODO. Hopefully, I'll remember to do the TODO in the next PR; it should be simple, but I want to avoid merge conflicts.